Source/Utility/UDim4.lua

```diff
@@ -1,34 +1,124 @@
-local UDim4 = {}
-UDim4.__index = UDim4
+-- @author MemoryPenguin
+-- @author Narrev
+-- UDim4 Object
 
-function UDim4.new(x, y)
-	assert(x and y, "Must supply x and y values.")
-	
-	local self = setmetatable({
-		X = x;
-		Y = y;
-	}, UDim4)
-	
-	return self
+local newVector2 = Vector2.new
+local newUDim2 = UDim2.new
+local IsA = game.IsA
+local SizePos = {
+	Size = "AbsoluteSize";
+	Position = "AbsolutePosition";
+}
+
+local function Connect(Event, func)
+	Connect = Event.Connect
+	Connect(Event, func)
+end
+
+local function Disconnect(Event)
+	Disconnect = Event.Disconnect
+	Disconnect(Event)
 end
 
-function UDim4.FromCoordinates(...)
-	return UDim4.new(
-		UDim2.new(select(1, ...)),
-		UDim2.new(select(5, ...))
+local function ToUDim2(self, GuiObject)
+	if not GuiObject.Parent or not IsA(GuiObject.Parent, "GuiBase2d") then
+		return newVector2(0, 0)
```

Comment:
This will cause anything that assumes this function returns a UDim2 to error when it tries to index `Scale` or `Offset` of a number. Return a zero-value UDim2 instead.